src/cache/cache_set.py

```diff
@@ -0,0 +1,275 @@
+        Update PLRU state when a way is accessed. If left child is accessed, parent bit is cleared.
+        If right child is accessed, parent bit is set.
+        """
+        if way_index >= self.num_ways:
+            raise ValueError(
+                f"Way index {way_index} out of range for {self.num_ways}-way set"
+            )
+
+        # PLRU bits are updated with a bottom up tree traversal
+        leaf_node = way_index + (
+            self.num_ways - 1
+        )  # Convert way index to a leaf_node number
+        node = self.__get_parent(
+            leaf_node
+        )  # Use leaf node to get first internal node (i.e. PLRU bit)
+
+        if way_index % 2 == 0:  # Left child accessed
+            self.state &= ~(1 << node)  # Clear node's bit
+        else:  # Right child accessed
+            self.state |= 1 << node  # Set node's bit
+
+        while node > 0:
+            parent = self.__get_parent(node)
+            if node == 2 * parent + 1:  # Current node is left child
+                self.state &= ~(1 << parent)  # Clear parent's bit
+            else:  # Current node is right child
+                self.state |= 1 << parent  # Set parent's bit
+            node = parent
+
+    def __get_plru_victim(self) -> int:
+        """
+        Find the way to replace according to PLRU policy.
+        Returns:
+            - Way index of victim line
+        """
+        node = 0
+        for _ in range(self.tree_levels):
+            if self.state & (1 << node):
+                node = 2 * node + 1  # Go left
+            else:
+                node = 2 * node + 2  # Go right
+        # Calculate leaf node index based on number of ways
+        return node - (self.num_ways - 1)
+
+    def __find_line(self, tag: int) -> Tuple[Optional[CacheLine], Optional[int]]:
+        """
+        Private Search Method: Find a cache line by tag.
+            Returns (line, way_index) or (None, None)
+        """
+        for i, line in enumerate(self.ways):
+            if not line.is_invalid() and line.tag == tag:
+                return line, i
+        return None, None
+
+    def read(self, tag: int) -> bool:
+        """
+        Simulates a read request to a cache set. If tag is found returns HIT, else MISS.
+        Upon a valid access plru bits are updated.
+        """
+        line, way_index = self.__find_line(tag)
+        if line is not None:  # Hit
+            # Update PLRU state bits on hit
+            self.__update_plru(way_index)
+            return True
+        return False  # Miss
+
+    def write(self, tag: int) -> bool:
+        """
+        Simulates a cache write request to a set. Write allocate policy is employed so any
+        write misses require cache to allocate line.
+        If tag is found returns HIT, else MISS.
+        """
+        line, way_index = self.__find_line(tag)
+        if line is not None:  # Hit
+            self.__update_plru(way_index)
+
+            return True
+        return False  # Miss
```

Comment:
I added a TODO comment. I think cache writes should return a MISS so the controller can issue a bus operation to simulate the action of filling a cache line from DRAM. Once it has the (figurative) data, it can issue an allocation to the cache set. This also allows other Caches to snoop on the bus operation. If it was handled internally by the cache set no bus operation would be simulated.